File: sections/11_intervals_sweep_line/problems/a_peak_active/solution.py
"""Reference solution for A. Peak Active."""

import sys


def solve_case(intervals: list[tuple[int, int]]) -> int:
    events: list[tuple[int, int]] = []
    for l, r in intervals:
        events.append((l, 1))
        events.append((r, -1))
    events.sort(key=lambda e: (e[0], e[1]))

    active = 0
    best = 0
    for _, delta in events:
        active += delta
        best = max(best, active)
    return best


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    t = data[0]
    idx = 1
    out: list[str] = []
    for _ in range(t):
        n = data[idx]
        idx += 1
        intervals: list[tuple[int, int]] = []
        for _ in range(n):
            l = data[idx]
            r = data[idx + 1]
            idx += 2
            intervals.append((l, r))
        out.append(str(solve_case(intervals)))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

File: sections/11_intervals_sweep_line/problems/a_peak_active/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int solve_case(const vector<pair<long long, long long>> &intervals) {
    vector<pair<long long, int>> events;
    events.reserve(intervals.size() * 2);
    for (auto [l, r] : intervals) {
        events.push_back({l, 1});
        events.push_back({r, -1});
    }
    sort(events.begin(), events.end(), [](const auto &a, const auto &b) {
        if (a.first != b.first) return a.first < b.first;
        return a.second < b.second;
    });

    int active = 0;
    int best = 0;
    for (auto [x, delta] : events) {
        (void)x;
        active += delta;
        best = max(best, active);
    }
    return best;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        int n;
        cin >> n;
        vector<pair<long long, long long>> intervals(n);
        for (auto &[l, r] : intervals) cin >> l >> r;
        cout << solve_case(intervals) << '\n';
    }

    return 0;
}

